### PR TITLE
Adding MPIRunner for mpi providers

### PR DESCRIPTION
**Pull request by sshudler**
This is a simple functionality to set a resource manager runner (Slurm, mpirun, etc.) for MPI providers. The immediate use for this feature will be in smoke tests where we need to run distributed code. This PR replaces #22699 which is closed due to some mistakes (had to rebase the branch before PR).

### CODE REVIEW

**Comment by jjellio** [comment]
Worth keeping in mind,  if you are cross compiling (or compiling with something like Cray), even a `serial` binary may need to be launched with the appropriate 'launcher' (using `-np 1`).  Be aware, (if the binary linked MPI), that even if the front-end arch == back-end arch (the binary will run) - it may still need to be launched by the parallel launcher. I hit this in the PNetCDF case, where the default behavior is to launch serial tests directly ... that assumption just doesn't hold up in most of the environments I'm in. (also didn't help that configure picked mpiexec over srun!)

**Comment by sshudler** [comment]
@jjellio That's a good point! The idea behind MPIRunner was to have a very simple way to use the resource manager to run MPI code, so it only checks for either `srun`, `mpiexec`, or `mpirun` and uses `which` to do that (similar to #22809). Perhaps it also makes sense to add `jsrun` and `aprun` - but we didn't want to make the MPIRunner overly complicated. You're right about using either `-n NP` or `-n 1` to run the test code if the package depends on an `mpi` provider and was linked against MPI libraries. But this is something every package will select by itself. MPIRunner is initialized in the `mpi` provider directly to be used by other packages in different scenarios, so setting up flags in it might not be good.

**Comment by jjellio** [comment]
It would be good to have a common place / tooling for determing the launcher

It would also be really good to have a common set of 'good flags'. Each package is doing this now... ugh. I've got to draft the PR to get HDF5 doing the same mojo PNetCDF does.

I take it, with your PR, the package would inherit from this class, then get some API to ask for the launcher?

Something that is *hard* is that you may have mpiexec in your path, and srun too. But you don't know which is the right one.  Sandia has a cluster that uses SLURM to manage jobs, but mpirun for launches.  We also had a system that used slurm for jobs and aprun for launches.  It's kinda bonkers.

IMO, it would be more robust to have some common variant. E.g., mpiexec=srun|mpirun|mpiexec|aprun|jsrun, and the ones in the list represent ones you support. It effectively just calls `which spec.variants.[mpiexec]` or something like that. You could also have a common variant for the flags (so the user could over ride).  Your class could then provide the API to gather that info. As is, everyone's doing their own thing.

I've ended up with replicated code across packages to determine the launcher + flags. I could see how this would atleast move that to a common place

**Comment by sshudler** [comment]
Having a common place for this functionality makes a lot of sense. I know that the packages `slate`, `strumpack`, and `superlu-dist` need it for sure.

MPIRunner is initialized as a member in every MPI provider, so the package doesn't inherit from it but just uses it like that: `self.spec['mpi'].runner`. The runner constructor basically selects the variant of the runner (tries `srun` first, then `mpiexec`, and so on). Adding default flags as a property will be good, and then each individual package could override them and use its own flags.

So how about I'll make these changes in the MPIRunner and the code to select the manager variant will be something like you used in PNetCDF, basically:
`mpiexecs = ['srun', 'aprun', 'jsrun']`
`mpiexecs_flags = ...`
And then for look for `['mpiexec', 'mpirun']` in the directory passed by the MPI provider?
The idea behind this last part is that `mpiexec` or `mpirun` could be in a directory which is specific to that particular MPI provider.

**Comment by jjellio** [comment]
I'd just be careful.  On Cray's (and really any platform using a job manager), you'll have an `mpiexec` in the MPI directory, but it's never the right choice. That's why I like the idea of a common variant for MPI packages. Then the user can just say, 'mpiexec=srun' in their spec and the ambiguity goes way.  The only constraint you really impose is that `which srun` should work... but if that doesn't the work, the user has problems. (I.e., they may need to have `slurm` in the MPI package's modules)

I guess one deal if the package is inherited is that you could apply a uniform variant - I'm guessing if it's in the MPI provider that would not allow a package to have the variant.

I'm afraid if you make it where each package needs to add a variant (or some means to know the name), you'll end up with the wild west of options. E.g., package X will have a bunch of `variants, +mpiexec, +mpirun +srun, +aprun`, .... others may have `launcher=` with `values('srun', 'mpirun')` (whatever that packager is familiar with).

It would be cleaner if the choice of launchers and means to provide flags was imposed uniformly. I think a simple package they could inherit from would do this - and call inherited methods (I guess - don't know python OO that well).  Then MPI projects could simply inherit, get the variants by default, and provide an API to return the either user flags or some defaults.

Don't take my word on things... I'm somewhat of a spack noob (but getting up to speed). I also highly value consistency. If any package supports MPI, and I can always say, `mpiexec=foo`, that's pretty nice from a maintenance perspective.

**Comment by sshudler** [comment]
I just realized the advantage in implementing it as a variant with a class you can derive from - basically, you can then adjust the runner command in the spec depending on where and how you install the packages. There might be two issues I think. One is that packages usually derive from classes that represent build systems (e.g., `AutotoolsPackage`) or just plain `Package` class. I'm not entirely sure how deriving from just a utility class such as MPIRunner fits into Spack's design philosophy. The other thing is that some MPI packages already have variants for schedulers. For example, `openmpi` has the `schedulers` variants whose value can be also 'slurm', so it might get confusing. But I'm actually a beginner in the world of Spack so I'm not entirely sure myself what would be a better approach.

@tldahlgren @becker33 What do you think?

**Comment by jjellio** [comment]
I'm mostly thinking aloud (in the covid work from home sense).

If the class inheritance system would work, it seems like it becomes quite elegant. Uniform variant options can be presented, while functionality (in the sense of an API) become directly usable w/out having to do anything to an MPI provider (or the virtual package itself).

I guess a different way to do it would be have MPI providers inherit instead, and then it kinda matches the idea you have to expose info through spec['mpi']. 

That said, packages derive from classes for more than build_systems. E.g., Cuda and ROCm.

The idea of packages in general that want MPI to derive from an MPI class seems kinda nice. It would really make sorting out launchers/options cleaner, since the details get hidden in exactly one place. Moreover, dependencies would be enforced. E.g., if someone wanted SLURM, then that magic class could impose `depends_on(slurm, type=run)`. It would entail the user having SLURM as a spack package (rather built or external).  The `depends_on` bit would eliminate the challenges of `which srun`

It's definitely worth chatting about with the SPACK guys (and I'm happy to take part if you want additional feedback)

**Comment by skosukhin** [comment]
I think that the `openmpi` implementation should somehow account for the `legacylaunchers` variant (see also #21961 where I try to make the variant more straightforward):
1. `mpiexec` might be still more preferable when `+legacylaunchers`;
2. `mpiexec` is a wrong fallback when `~legacylaunchers` (`orterun` could be an option though).

Also, a use case that would be nice to consider. Having an `mpi` provider built with `slurm` support and the correct (which is also not necessarily the case) `srun` in the `PATH` does not mean that I want to use it when installing a package with Spack. For example, on one of our systems, the users are supposed to launch their real heavy applications with `srun` but they are not discouraged from running small MPI programs on the login nodes using `mpiexec`. This is very helpful when we install a package that runs small MPI programs at the configure time (e.g. to make sure that the provided MPI library doesn't suffer a particular bug) or at the `make check` stage. Yes, we can still use `srun` but we would like to avoid it in this case because we don't want to wait in the queue when we don't have to and because running `srun` without additional custom arguments is impossible (see #22824).

**Comment by jjellio** [comment]
I don't understand the latter statement. 
>Yes, we can still use srun but we would like to avoid it in this case because we don't want to wait in the queue when we don't have to and because running srun without additional custom arguments is impossible (see #22824).

This is exactly why I want something standard that can be applied top-down via variants: e.g., 
```
spack install something_depends_on_mpi +blah~bluh \
           mpiexec="$(which srun)" \
           mpiexec_flags="-A some_account -C some_constraint -p some_partition -n"
```

There's nothing technically blocking you from using arbitrary args with `srun` - the blocking is that there is no way to communicate `flags` in a sane way. Each package has to do their own thing, which isn't nice.

I advocate for not working about figuring out what the user wants, and instead providing an interface so the user can tell you what they want.  It's the wild west when you start inferring things anyway. E.g., the policy of choose the first executable found in `[ srun, aprun, mpiexec ]` (And then impose reasonable default flags).  That's all pure speculation and breaks when you want to run on a production environment that wants an account (or other arbitrary data supplied).

Also, be careful with MPI' launchers - OpenMPI usually builds *support* for a launcher (typically via PMI), they don't actually provide SRUN or anything (in days of yore they provided LAM boot - but that ship sailed decades ago). MPICH does provide some launchers (hydra and pals) - but you usually build support for some managers (slurm, pbs, etc..)

IMO, if you can define a clean interface to expose the idea of "what should be used to execute a binary" and "what flags does it require" - Clearly explaining that the flags should *end* with the argument before NP.

If you want to get more complicated, in Trilinos we have (in CMake)
```
MPI_EXEC_NUMPROCS_FLAG=-np

(The command-line option just before the number of processes to use <NP>. The default value is based on the name of ${MPI_EXEC}, for example, which is -np for OpenMPI.)

-D MPI_EXEC_PRE_NUMPROCS_FLAGS="arg1;arg2;...;argn"
(Other command-line arguments that must come before the numprocs argument. The default is empty "".)

-D MPI_EXEC_POST_NUMPROCS_FLAGS="arg1;arg2;...;argn"
```

For the most part, I've never needed pre/post flags, because I can always organize my flags so that they end with `-n ` (Ready for the program to tack on the number of procs)

We use the prior CMake stuff to run on all manner of production machines (Nerscs, Serria, Summit, Trinity) and the horde of test machines as they appear.

IMO, the requisite information is:
* What is the command to run (srun, aprun, etc...)
* What args does it take (ending ready for NP)

On several platforms I'll end up writing a wrapper around one of the launchers - and apply special things before the launch (we have to do this with things like device-aware MPI, where it can be toggled on/off in our code and in MPI - and you need both in sync before launching)

I also wouldn't worry about the MPI packages - the cmake `build_type` top-level variant happily works among many packages. I kinda view this as a means to present another two top-level attributes that become available if you use MPI. The code in the class/methods just makes accessing / using that info more standard (I didn't learn till recently I could do `spec.variants['name'].value`

**Comment by tldahlgren** [comment]
@tgamblin @becker33 There is an interesting discussion going on here regarding `MPIRunner` and an alternative along the lines of `cuda.py` and `rocm.py`. 

My preference is for the latter since I too like the idea of inheriting an interface for dealing with the runners.  It provides a single place for collecting variants, known conflicts, etc. that can be shared.  But I am not familiar with the corner cases.

**Comment by skosukhin** [comment]
@jjellio you don't have to convince me of anything here (sorry if I'm wrong but to me, it looked like you tried to) because I was not arguing against your remarks and suggestions (and my comment wasn't addressed to you in particular). I just wanted to add to the discussion a use-case that shows that the solution must be flexible and guessing is not really an option.

I totally agree that adding a variant could be an option. We could start with:
1. Add a base class to [build_systems](https://github.com/spack/spack/tree/develop/lib/spack/spack/build_systems) (e.g. see [CudaPackage](https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/cuda.py) and how it is used in [Opencv](https://github.com/spack/spack/blob/c4a83aa22cd8f08bd4c348b93941440c02da80be/var/spack/repos/builtin/packages/opencv/package.py#L5)):
    ```python
    from spack.directives import variant
    from spack.package import PackageBase
    
    
    def mpi_launch_values(x):
        """Pretty string describing possible values for mpi_launch
        """
        # The docstring above is used for the 'spack info' output.
        return x
    
    
    class MPIRunnerPackage(PackageBase):
        variant('mpi_launch', default='none', values=mpi_launch_values,
                description='MPI launcher command')
    ```
2. Register it in [pkgkit.py](https://github.com/spack/spack/blob/develop/lib/spack/spack/pkgkit.py)
3. Use it in, for example, [ParallelNetcdf](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/parallel-netcdf/package.py) as follows:
    ```python
    class ParallelNetcdf(AutotoolsPackage, MPIRunnerPackage):
        ...
        def configure_args(self):
            ...
            mpi_launch_cmd = self.spec.variants['mpi_launch'].value
            if mpi_launch_cmd != 'none':
                args.append('TESTMPIRUN={0} NP'.format(mpi_launch_cmd))
            ...
    ```
4. And install the package as follows:
    ```console
    spack install parallel-netcdf mpi_launch='srun -n'
    ```
    We can even set the default in `packages.yaml` for all packages:
    ```yaml
    packages:
      all:
        variants: mpi_launch='srun -n'
    ```
    and run simply:
    ```console
    spack install parallel-netcdf
    ```

There are at least two problems with this solution though:
1. The value of the `mpi_launch` affects the hash, which means that the following very similar commands will result in two different installations:
    ```console
    spack install parallel-netcdf mpi_launch='mpiexec -n'
    spack install parallel-netcdf mpi_launch='mpiexec  -n'
    ```
    I don't know whether it is possible to tell Spack to ignore `mpi_launch` when calculating the hash.
2. It would be nice to be able to set the value for `mpi_launch` only once and do not repeat it on the command line for every package that needs it (like in the example with `packages.yaml` above). However, the value (e.g. `srun -n`) might be correct for some `mpi` providers (e.g. `openmpi` with `schedulers=slurm`) and incorrect for other (e.g. `openmpi` without SLURM support). Therefore, some light guessing (unless specified explicitly on the command line) or additional flexibility in the configuration file would be still nice to have.

**Comment by sshudler** [comment]
@skosukhin I agree with this approach. The last point you raise - about the value of `mpi_launch` not being correct for some `mpi` providers is one of the reasons behind this PR. This is why the MPIRunner is assigned to `self.spec.runner` in each one of the providers. However, I think your approach for the variants could be tweaked a bit. How about:

```python
from spack.directives import variant
from spack.package import PackageBase


def mpi_launch_values(x):
    """Pretty string describing possible values for mpi_launch
    """
    # The docstring above is used for the 'spack info' output.
    return x


class MPIRunnerPackage(PackageBase):
    variant('mpi_launch', default='none', values=mpi_launch_values,
            description='MPI launcher command')

    # Basically, this function is used to adjust the value of 'mpi_launch' provided
    # by the user depending on the mpi provider in the spec
    def setup_dependent_package(self, module, dependent_spec):
        if 'opempi' in self.spec and 'schedulers=slurm' in self.spec:
            self.variants['mpi_launch'] = 'srun'
        # ...
```

**Comment by jjellio** [comment]
Atleast in Trilinos land, we've had success keeping the launcher and flags separate. (e.g., two options, `mpiexec=srun` and `mpiexec_flags=-n`). It's simple on some machines, but we support Windows, and Bill Gates loves spaces. So, splitting `launcher` on white space is never safe. Yet, when we interact with CMake, it wants `MPI_EXEC` and `MPI_EXEC_FLAGS` (or some spelling of that).

Imo, it's worth being aware that white spaces in paths are a thing outside the Linux community (and spack does support those systems)

One problem I've had with:
```
        if 'opempi' in self.spec and 'schedulers=slurm' in self.spec:
            self.variants['mpi_launch'] = 'srun'
        # ...
```
When I hacked things in HDF5/NetCDF/PNetCDF, is that `mvapich` has `process_managers=auto` by default - in this case you have no idea what's going on.  The user (i.e., ME) could annotate `mvapich` with `^slurm`, but that didn't always play out either. This comes up because `mvapich` is **external**, *so your definition of the spec is kinda ad-hoc. You may write it nicely, but joe-user on their first pony ride on NERSC Cori probably won't.*

In the case of class inheritance, a point comes down to the use of inherited API vs through the spec. E.g., `self.runner()` or  `self.spec.runner` .

I can't think of a draw back to one or the other.

**Comment by sshudler** [comment]
@jjellio I agree that it makes more sense to separate the launcher and the flags, something like that:

```python
variant('mpiexec', default='none', values=mpiexec_values,
        description='MPI exec command')
variant('mpiexec_flags', default='none', values=mpiexec_flags,
        description='MPI exec flags')
```

**Comment by jjellio** [comment]
@sshudler If you come up with something you want me to try out, just mention some instructions.

**Comment by sshudler** [comment]
Thanks! I'll work now on this change we discussed and update this PR

**Comment by Try2Code** [comment]
hi! I just wanted to emphasis on the danger of default values here: we have 
* different MPI vendors
* different build-variants
* different HPC-system demands

Using spack to guess on what the user might want is a very dangerous path to go. Coming with the best intentions it will for sure lead to strange behavior that will be fixed in another part of the code.

IMO users should be able to set their needs on the command line or in any other well known configuration area. I agree with @skosukhin that a new hash value might be an unwanted consequence but tbh I rather use a deterministic and explicit package manager than smart-looking guesswork. no offense, but a package manager should offer a clearly defined way of configuring the build/installation process. for me _guessing_ does not belong to that category

**Comment by skosukhin** [comment]
@sshudler the more I think about it, the less I like the solution that I suggested. Variants are not really meant for such things (especially, given the current limitations described in https://github.com/spack/spack/pull/22703#issuecomment-815093329). Also, since we get the MPI compiler wrappers from the `mpi` spec (e.g. `self.spec['mpi'].mpicc`) it would make sense to get the launch command in the same way (e.g. `self.spec['mpi'].mpi_launch`). Especially since the value should depend on the particular `mpi` provider.

I think that it would be better to reserve an environment variable `SPACK_MPI_LAUNCH` (the name and whether we need something like `SPACK_MPI_LAUNCH_ARGS` can be discussed separately) and say that every `mpi` provider must respect it in `setup_dependent_package`. For example, for `openmpi` it would be something like:
```python
def setup_dependent_package(self, module, dependent_spec):
    ...
    self.spec.mpi_launch = os.environ.get('SPACK_MPI_LAUNCH', None)
    # TODO: the following needs refactoring:
    if not self.spec.mpi_launch:
        if 'schedulers=slurm' in self.spec:
            if '+legacylaunchers' in self.spec:
                self.spec.mpi_launch='{0} -n'.format(self.spec.prefix.bin.mpiexec)
            elif 'slurm' in self.spec:
                self.spec.mpi_launch='{0} -n'.format(self.spec['slurm'].prefix.bin.srun)
            else:
                # This should cover the case when we are using an external `openmpi`,
                # which might mean (needs to be checked) that `slurm` is not in self.spec:
                srun = which('srun')
                if srun:
                    self.spec.mpi_launch='{0} -n'.format(srun.exe)
                else:
                    self.spec.mpi_launch='{0} -n'.format(self.spec.prefix.bin.orterun)
        else:
            self.spec.mpi_launch='{0} -n'.format(self.spec.prefix.bin.mpiexec)
    else:
        self.spec.mpi_launch='{0} -n'.format(self.spec.prefix.bin.orterun)
```

This way, the users are supposed to export `SPACK_MPI_LAUNCH` with whatever value they want before running `spack`. Once #22824 is implemented, they (we :) will be able to set the command they like for each `mpi` provider specifically or for all of them at once in `packages.yaml`. We can also think of introducing an additional entry for this in `config.yaml` but that seems redundant for now.

**Comment by jjellio** [comment]
Plus+1 for not guessing things ... if possible just expose necessary parameters and users can set them.

From a robustness perspective, it may be beneficial to define the 'flags' however you set them as :`Flags can be anything needed and you should place NP as a placeholder for the number of processes`. - I know that's annoying, but you never what a system is going to require.

Your tooling - parses either a variant, ENV, or config and populates some config info e.g., `MY_TESTCOMMAND_SERIAL = self.something.mpirun + ' ' + self.something.mpirun_flags(np=1)` and `MY_TESTCOMMAND_PARALLEL = ... self.something.mpirun_flags(np=6)` (this would enable setting up things like  PNetCDF or HDF5's testing) 

This would hide the details from packagers about where those values come from (so in the future if they shifted from ENVs to  config.yaml, packages just work)

From a Trilinos' perspective, I need to be able to get the command and flags (separately).  E.g., a Sierra launch can look like: `-c2;-g1;-r4;-brs;-p`  (ending with NP). To this end, we do support non-ending w/NP flags... hence we have `PRE_NP_FLAGS` `NP_FLAGS` and `POST_NP_FLAGS` (ugh!) - ugly details, but at some point on some system they've been used. 

If you can wrap the details of whatever sets these (variants, ENV, or config yaml) - then most users could default to setting `NP_FLAGS = -n`.

The user (package) interface would then have
```
fields (variables/functions)
self.something.mpiexec
self.something.mpiexec_pre_np_flags
self.something.mpiexec_post_np_flags
self.something.mpiexec_np_flags

any maybe some nice top-level function that people actually call
# the following just stitching together these options 
self.something.mpiexec_cmd(np=1) 
```
An alternative would be define 'flags' using mini-syntax, e.g., `FLAGS="whatever you want NP whatever else you want"` and document that NP will get replaced with ... (which still would still support `self.something.mpiexec_cmd(np=1)` )
**Edit to add: I am more an advocate of not using the NP mini-syntax (as it requires you to split it up)**

In practice, the user would usually call `self.something.mpiexec_cmd(np=1)` to get a full command ready-made.... but packages could also use those fields/properties/whatever to setup per-package build things (e.g., Trilinos)

**Comment by jjellio** [comment]
The more I think about this, the more I favor config options in a yaml.
1.  It makes the variable concise and located in one place. e.g.,
```yaml
confg.yaml:
  # type is a string
  mpiexec : default = None  (call it whatever! mpiexec seems like a poor choice)
  # type is string or list (but respect the list! - e.g., store it internally as [str] or the list provided)
  mpiexec_np_flags : default = None  
  mpiexec_pre_np_flags : default = None (Same type semantics as above)
  mpiexec_post_np_flags : default = None
```
2.  If I write a spack environment, these variables are cleanly visible (if I organize the file correctly) - not buried inside packages
3. Types! (and unpacking) - the expectation is that packages will want to consume these values. And possibly deal with them as sets of flags (e.g., configuring CMake / Ctest)
A practical example of a pit fall:  On Sierra/Summit I have some people that need to set 
    *    `MPI_EXEC=jsrun`, and `MPI_EXEC_NP_FLAGS=-M;'-gpu -mca ...';-r4;-g1;-c1;-brs -p`
    * The point is their flags do have white space
    * The flag needs to respect those quotes, e.g., Spectrum wants -M ' whatever you want want' 
    * if this were all one flat string, it would be hard to parse.
    * The prior config would have `mpiexec_np_flags : [ -M, -gpu -mca....(don't know if you need to quote it), -r4, -g1, -brs, -p]`

The api is the same as described prior. Convenience methods can `join` the lists - and you could even quote them (what happens is just details of your choosing)
E.g., returning `'srun' '-n'`.

The properties/fields would be lists internally - this would provide the safest method for handling bonker flags the user may throw.  Ultimately, I'd be feeding these flags into CMake variables.  But It think the general issue presented above is pretty important

**Comment by sshudler** [comment]
Originally, the idea for the MPI runner was to have a simple way to run smoke tests rather than a full-fledged mechanism that mirrors a resource manager. Some guessing or querying through `which` simplifies things. But I agree that we should expose parameters so that users can set values for specific environments.

After giving it some thought, I agree with @skosukhin that variants are problematic - they change the hash of the spec and it can make build caches unusable. I won't be able to use something prebuilt, which fits my system hw/compiler, just because the resource manager is configured differently...

@jjellio I think it's good to separate the flags from the mpiexec runner command, and also have separate `mpiexec_np_flags`, `mpiexec_pre_np_flags`, and `mpiexec_post_np_flags` parameters.

Choosing between ENV variables and `config.yaml` I think I tend to prefer `config.yaml`, but as long as it allows customizing the launcher command/flags per MPI provider in the same way as described in #22824:
```python
confg.yaml:
  mpiexec:
  mpiexec_np_flags:
  mpiexec_pre_np_flags:
  # ...
  openmpi_cfg:
    mpiexec:
    mpiexec_np_flags:
    # ...
```

**Comment by prckent** [comment]
Speaking as a random user, use of config.yaml seems problematic and error prone since different packages may have different launch requirements. While it is easy to adjust environment variables from typical shell scripts, the same can not be said of yaml. i.e. Important to maintain an environment variables route and catch any incompatible specifications if possible.

**Comment by jjellio** [comment]
@psakievich

**Comment by wspear** [comment]
A complete implementation of a launch management system like this looks necessary to me. In E4S 23.02 we found that on Crusher all sequential executables need to be launched with `srun` or we get a failure like

`MPICH ERROR: Unable to use a NIC_POLICY of 'NUMA'. Rank 0 is not confined to a single NUMA node.  There are 4 numa_nodes detected (rc=0).`

Currently there is no clear path to successfully running spack smoke tests affected by this issue on an existing Spack deployment.

In theory that would be any spack test that launches a sequential executable, but we've observed it so far in:

- darshan-runtime
- warpx
- pumi
- amrex

**Comment by alalazo** [comment]
Closing, since there has been no activity in a while, and due to the number of conflicts. Feel free to reopen if you want to continue work on this.
